cryptographic_estimators/UOVEstimator/uov_algorithm.py

```diff
@@ -0,0 +1,79 @@
+# ****************************************************************************
+# Copyright 2023 Technology Innovation Institute
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# ****************************************************************************
+
+
+from ..base_algorithm import BaseAlgorithm
+from .uov_problem import UOVProblem
+from sage.arith.misc import is_prime_power
+
+
+class UOVAlgorithm(BaseAlgorithm):
+    def __init__(self, problem: UOVProblem, **kwargs):
+        """
+        Base class for UOV algorithms complexity estimator
+
+        INPUT:
+
+        - ``problem`` -- UOVProblem object including all necessary parameters
+        - ``w`` -- linear algebra constant (default: 2)
+        - ``h`` -- external hybridization parameter (default: 0)
+        - ``memory_access`` -- specifies the memory access cost model (default: 0, choices: 0 - constant, 1 - logarithmic, 2 - square-root, 3 - cube-root or deploy custom function which takes as input the logarithm of the total memory usage)
+        - ``complexity_type`` -- complexity type to consider (0: estimate, 1: tilde O complexity, default: 0)
+        - ``bit_complexities`` -- determines if complexity is given in bit operations or basic operations (default 1: in bit)
+
+        """
+        super(UOVAlgorithm, self).__init__(problem, **kwargs)
+
+        h = kwargs.get("h", 0)
+        w = kwargs.get("w", 2.81)
+        n = self.problem.nvariables()
+        m = self.problem.npolynomials()
+        q = self.problem.order_of_the_field()
```

Comment:
since those parameters are not used until they are stored in self._n, ... I think we can delete n, m, q, and store directly in self._n ... but see next comment